packages/shared-crypto/lib/keyring.ts

```diff
@@ -0,0 +1,314 @@
+MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+GNU Lesser General Public License for more details.
+
+You should have received a copy of the GNU Lesser General Public License
+along with the library. If not, see <http://www.gnu.org/licenses/>.
+*/
+
+import { AddressHash } from '@alephium/shared'
+import { addressToGroup, bs58, ExplorerProvider, sign, TOTAL_NUMBER_OF_GROUPS, transactionSign } from '@alephium/web3'
+import * as bip32 from 'bip32'
+import * as bip39 from 'bip39'
+import blake from 'blakejs'
+
+import { findNextAvailableAddressIndex, isAddressIndexValid } from './address'
+import { decryptMnemonic, MnemonicLength } from './wallet'
+
+export type NonSensitiveAddressData = {
+  hash: AddressHash
+  index: number
+  publicKey: string
+}
+
+type SensitiveAddressData = NonSensitiveAddressData & {
+  privateKey: Buffer
+}
+
+type GenerateAddressProps = {
+  group?: number
+  addressIndex?: number
+  skipAddressIndexes?: number[]
+}
+
+class Keyring {
+  private hdPath = "m/44'/1234'/0'/0"
+  private root: bip32.BIP32Interface | null
+  private addresses: SensitiveAddressData[]
+
+  constructor() {
+    this.addresses = []
+    this.root = null
+  }
+
+  // PUBLIC METHODS
+
+  public clearCachedSecrets = () => {
+    this.addresses = []
+    this.root = null
+  }
+
+  public generateRandomMnemonic = (mnemonicLength: MnemonicLength = 24): Buffer => {
+    const strength = mnemonicLength === 24 ? 256 : 128
+    const mnemonic = Buffer.from(bip39.generateMnemonic(strength))
+
+    this._initFromMnemonic(mnemonic, '')
+
+    return mnemonic
+  }
+
+  public importMnemonicString = (mnemonic: string) => {
+    if (!mnemonic) throw new Error('Keyring: Cannot import mnemonic, mnemonic not provided')
+
+    this.clearCachedSecrets()
+    this._initFromMnemonic(Buffer.from(mnemonic), '')
+
+    mnemonic = ''
+  }
+
+  public initFromEncryptedMnemonic = (encryptedMnemonic: string, password: string, passphrase: string) => {
+    let decryptedMnemonic: Buffer | null = decryptMnemonic(encryptedMnemonic, password).decryptedMnemonic
+    this.clearCachedSecrets()
+    this._initFromMnemonic(decryptedMnemonic, passphrase)
+
+    encryptedMnemonic = ''
+    password = ''
+    passphrase = ''
+    decryptedMnemonic = null
+  }
+
+  public generateAndCacheAddress = (props: GenerateAddressProps): NonSensitiveAddressData =>
+    this._getNonSensitiveAddressData(this._generateAndCacheAddress(props))
+
+  public signTransaction = (txId: string, addressHash: AddressHash): string =>
+    transactionSign(txId, this.exportPrivateKeyOfAddress(addressHash).toString('hex'))
+
+  public signMessage = (message: string, addressHash: AddressHash): string =>
+    sign(message, this.exportPrivateKeyOfAddress(addressHash).toString('hex'))
+
+  public exportPrivateKeyOfAddress = (addressHash: AddressHash): SensitiveAddressData['privateKey'] =>
+    this._getAddress(addressHash).privateKey
+
+  public discoverAndCacheActiveAddresses = async (
```

Comment:
No changes were made to this algorithm, it was just copied over so that it can use the secrets that should not leave the keyring